paymentsheet/src/test/java/com/stripe/android/paymentsheet/elements/TextFieldElementTest.kt

```diff
@@ -0,0 +1,273 @@
+    }
+
+    @Test
+    fun `Verify filter called on value change`() {
+        val configErrorBasedOnFocus = TestConfigErrorBasedOnFocus()
+        val textFieldElementErrorBasedOnFocus = TextFieldElement(configErrorBasedOnFocus)
+
+        var visibleError = false
+        textFieldElementErrorBasedOnFocus.visibleError
+            .observeForever {
+                visibleError = it
+            }
+
+        textFieldElementErrorBasedOnFocus.onFocusChange(false)
+        assertThat(visibleError).isEqualTo(false)
+
+        textFieldElementErrorBasedOnFocus.onFocusChange(true)
+        shadowOf(getMainLooper()).idle()
+        assertThat(visibleError).isEqualTo(true)
+    }
+
+    @Test
+    fun `Verify filter is called to set the input value`() {
+        val numberConfigFilter = TestConfigFilter()
+        val textFieldElement = TextFieldElement(numberConfigFilter)
+
+        var inputValue = ""
+        textFieldElement.input
+            .observeForever {
+                inputValue = it
+            }
+
+        textFieldElement.onValueChange("1a2b3c4d")
+        shadowOf(getMainLooper()).idle()
+        assertThat(inputValue).isEqualTo("1234")
+    }
+
+    private class TestConfig : TextFieldConfigInterface {
+        override val debugLabel = "debugLabel"
+        override val label: Int = R.string.address_label_name
+        override val keyboard: KeyboardType = KeyboardType.Ascii
+
+        var fakeShouldShowError = false
+        var fakeElementState: TextFieldElementState =
+            TextFieldElementState.TextFieldElementStateValid()
+
+        override fun determineState(paramFormatted: String?): TextFieldElementState =
+            fakeElementState
+
+        override fun shouldShowError(
+            elementState: TextFieldElementState,
+            hasFocus: Boolean
+        ) = fakeShouldShowError
+
+        override fun filter(userTyped: String): String = userTyped
+    }
+
+    private class TestConfigErrorBasedOnFocus : TextFieldConfigInterface {
+        override val debugLabel = "debugLabel"
+        override val label: Int = R.string.address_label_name
+        override val keyboard: KeyboardType = KeyboardType.Ascii
+
+        var fakeElementState: TextFieldElementState =
+            TextFieldElementState.TextFieldElementStateValid()
+
+        override fun determineState(paramFormatted: String?): TextFieldElementState =
+            fakeElementState
+
+        override fun shouldShowError(
+            elementState: TextFieldElementState,
+            hasFocus: Boolean
+        ) = hasFocus
+
+        override fun filter(userTyped: String): String = userTyped
+    }
+
+    private class TestConfigFilter : TextFieldConfigInterface {
+        override val debugLabel = "debugLabel"
+        override val label: Int = R.string.address_label_name
+        override val keyboard: KeyboardType = KeyboardType.Ascii
+
+        override fun determineState(paramFormatted: String?): TextFieldElementState =
```

Comment:
What does `paramFormatted` represent? When is it null?

Reply:
wrongly named, updated.